File: libgraffiti/src/renderer.rs
// turns render tree into series of skia calls
//
// maybe we can get rid of skia one day or make it an optional feature/backend
// but it's not a priority right now and there are no good alternatives
// at the moment anyway (clip radii, filters)
//
// notes:
// - render tree is just a slice of `RenderEdge`(s) which means it should be prefetch-friendly
//   and the `ContainerStyle` doesn't need to be <64b
// - incrementality will be handled elsewhere, this should be as simple and stateless as possible
//   except of maybe using some LRU caches for managing mid-lived resources, etc.

use skia_safe::textlayout::Paragraph;
use skia_safe::{
    gpu::{gl::FramebufferInfo, BackendRenderTarget, DirectContext},
    Canvas, ColorType, Paint, RRect, Surface,
};
use skia_safe::{ClipOp, MaskFilter};
use std::ops::Deref;

// for now we just re-export some skia primitives
pub use skia_safe::{Color, Matrix, Point, Rect};

pub enum RenderEdge<P: Deref<Target = Paragraph>> {
    OpenContainer(Rect, ContainerStyle),
    CloseContainer,
    Text(Rect, P),
}

#[derive(Debug, Clone, Copy, Default)]
pub struct ContainerStyle {
    pub transform: Option<Matrix>,
    pub opacity: Option<f32>,
    pub border_radii: Option<[f32; 4]>,
    pub shadow: Option<Shadow>,
    pub outline: Option<Outline>,
    pub clip: bool,
    pub bg_color: Option<Color>,
    // pub images/gradients: Vec<?>
    pub border: Option<Border>,
}

#[derive(Debug, Clone, Copy)]
pub struct Shadow(pub (f32, f32), pub f32, pub f32, pub Color);

#[derive(Debug, Clone, Copy)]
pub struct Outline(pub f32, pub Option<StrokeStyle>, pub Color);

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum StrokeStyle {
    Solid,
    Dashed,
    Dotted,
}

#[derive(Debug, Clone, Copy)]
pub struct Border {
    top: BorderSide,
    right: BorderSide,
    bottom: BorderSide,
    left: BorderSide,
}

#[derive(Debug, Clone, Copy)]
pub struct BorderSide(pub f32, pub Option<StrokeStyle>, pub Color);

pub struct Renderer {
    gr_ctx: DirectContext,
    surface: Surface,
}

struct RenderContext<'a> {
    canvas: &'a mut Canvas,
}

enum Shape {
    Rect(Rect),
    RRect(RRect),
}

impl Renderer {
    pub fn new(size: (i32, i32)) -> Self {
        let mut gr_ctx = DirectContext::new_gl(None, None).unwrap();
        let surface = Self::create_surface(&mut gr_ctx, size);

        Self { gr_ctx, surface }
    }

    pub fn render<P: Deref<Target = Paragraph>>(&mut self, render_tree: &[RenderEdge<P>]) {
        let canvas = self.surface.canvas();
        canvas.clear(Color::WHITE);

        let mut ctx = RenderContext { canvas };

        for edge in render_tree {
            ctx.draw_edge(edge);
        }

        self.surface.flush();
    }

    pub fn resize(&mut self, size: (i32, i32)) {
        self.surface = Self::create_surface(&mut self.gr_ctx, size);
    }

    fn create_surface(gr_ctx: &mut DirectContext, size: (i32, i32)) -> Surface {
        let target = BackendRenderTarget::new_gl(
            size,
            None,
            8,
            FramebufferInfo {
                fboid: 0,
                format: skia_safe::gpu::gl::Format::RGBA8.into(),
            },
        );

        Surface::from_backend_render_target(
            gr_ctx,
            &target,
            skia_safe::gpu::SurfaceOrigin::BottomLeft,
            ColorType::RGBA8888,
            None,
            None,
        )
        .unwrap()
    }
}

impl RenderContext<'_> {
    fn draw_edge<P: Deref<Target = Paragraph>>(&mut self, edge: &RenderEdge<P>) {
        match edge {
            RenderEdge::OpenContainer(rect, style) => self.open_container(*rect, style),
            RenderEdge::CloseContainer => self.close_container(),
            RenderEdge::Text(rect, paragraph) => self.draw_text(*rect, paragraph),
        }
    }

    fn open_container(&mut self, rect: Rect, style: &ContainerStyle) {
        self.canvas.save();

        let shape = match &style.border_radii {
            None => Shape::Rect(rect),
            Some(radii) => {
                let mut rrect = RRect::default();
                let &[a, b, c, d] = radii;
                rrect.set_rect_radii(rect, &[(a, a).into(), (b, b).into(), (c, c).into(), (d, d).into()]);

                Shape::RRect(rrect)
            }
        };

        if let Some(matrix) = &style.transform {
            self.canvas.concat(matrix);
        }

        if let Some(opacity) = style.opacity {
            // TODO
        }

        if let Some(shadow) = &style.shadow {
            self.draw_shadow(rect, shadow);
        }

        if let Some(outline) = &style.outline {
            self.draw_outline(rect, outline);
        }

        if style.clip {
            // TODO: subtract borders?
            self.clip_shape(&shape, ClipOp::Intersect, style.transform.is_some());
        }

        if let Some(color) = style.bg_color {
            self.draw_bg_color(&shape, color);
        }

        // TODO: image(s)

        // TODO: scroll
        self.canvas.translate((rect.x(), rect.y()));
    }

    fn close_container(&mut self) {
        //let (border,) = ctx.stack.pop().unwrap();

        // if let Some(border) = ctx.stack.pop {
        //     self.draw_border(border_rect, border_radii, border)
        // }

        self.canvas.restore();
    }

    fn draw_text(&mut self, rect: Rect, paragraph: &Paragraph) {
        paragraph.paint(self.canvas, Point::new(rect.x(), rect.y()));
    }

    fn draw_outline(&mut self, rect: Rect, outline: &Outline) {
        let &Outline(width, style, color) = outline;

        let mut paint = Paint::default();
        paint.set_stroke(true);
        paint.set_stroke_width(width);
        paint.set_color(color);

        // TODO: stroke styles

        let d = width / 2.;

        self.canvas.draw_rect(rect.with_outset((d, d)), &paint);
    }

    fn draw_bg_color(&mut self, shape: &Shape, color: Color) {
        let mut paint = Paint::default();
        paint.set_color(color);

        self.draw_shape(shape, &paint);
    }

    // TODO: radii
    fn draw_shadow(&mut self, rect: Rect, shadow: &Shadow) {
        let &Shadow(offset, blur, spread, color) = shadow;

        let mut paint = Paint::default();
        paint.set_color(color);
        paint.set_mask_filter(MaskFilter::blur(skia_safe::BlurStyle::Outer, sigma(blur), false));

        // TODO: fix negative spread (visible with transparent background)
        //       maybe draw those with inverse clip?
        self.canvas
            .draw_rect(rect.with_offset(offset).with_outset((spread, spread)), &paint);
    }

    fn draw_shape(&mut self, shape: &Shape, paint: &Paint) {
        match shape {
            Shape::Rect(rect) => self.canvas.draw_rect(rect, paint),
            Shape::RRect(rrect) => self.canvas.draw_rrect(rrect, paint),
        };
    }

    fn clip_shape(&mut self, shape: &Shape, op: ClipOp, antialias: bool) {
        match shape {
            Shape::Rect(rect) => self.canvas.clip_rect(rect, op, antialias),
            Shape::RRect(rrect) => self.canvas.clip_rrect(rrect, op, antialias),
        };
    }
}

fn sigma(radius: f32) -> f32 {
    // (1 / sqrt(3))
    0.5773502692 * radius + 0.5
}

File: libgraffiti/src/renderer.zig
const std = @import("std");
const nvg = @import("nanovg");
const Document = @import("dom/dom.zig").Document;
const NodeId = @import("dom/dom.zig").NodeId;
const Style = @import("style.zig").Style;
const Color = @import("style.zig").Color;
const TRANSPARENT = @import("style.zig").TRANSPARENT;
const LayoutNode = @import("layout/layout.zig").LayoutNode;
const calculate = @import("layout/layout.zig").calculate;

const Shape = union(enum) { rect: Rect, rrect: RRect };

const Rect = struct { x: f32 = 0, y: f32 = 0, w: f32 = 0, h: f32 = 0 };

const RRect = struct { rect: Rect, radii: [4]f32 };

pub const Renderer = struct {
    allocator: std.mem.Allocator,
    vg: nvg,

    const Self = @This();

    pub fn init(allocator: std.mem.Allocator) anyerror!Self {
        const vg = try nvg.gl.init(allocator, .{
            .antialias = true,
            .stencil_strokes = false,
            .debug = true,
        });

        const font = @embedFile("../nanovg-zig/examples/Roboto-Regular.ttf");
        _ = vg.createFontMem("sans", font);

        return Self{ .allocator = allocator, .vg = vg };
    }

    pub fn deinit(self: *Self) void {
        self.vg.deinit();
    }

    pub fn render(self: *Self, document: *Document, w: f32, h: f32) void {
        std.debug.print("TODO: style is empty for now\n", .{});

        self.vg.reset();
        self.vg.beginFrame(w, h, 1.0);
        defer self.vg.endFrame();

        // layout-all
        var layout_nodes = self.allocator.alloc(LayoutNode, document.nodes.list.len) catch @panic("oom");
        defer self.allocator.free(layout_nodes);
        for (layout_nodes) |*l, n| {
            const node = document.nodeById(n);

            l.* = switch (node.nodeType()) {
                .element => .{ .style = .{} },
                .text => .{ .style = .{ .display = .@"inline" }, .text = node.text().?.data },
                .document => .{ .style = .{} },
                .comment, .document_fragment => .{ .style = .{} },
            };

            l.first_child = if (node.first_child) |ch| &layout_nodes[ch.id] else null;
            l.next = if (node.next_sibling) |ne| &layout_nodes[ne.id] else null;
        }
        calculate(&layout_nodes[document.node.id], .{ .width = w, .height = h });

        var ctx = RenderContext{ .vg = &self.vg };

        // white bg
        // TODO: https://github.com/ziglang/zig/issues/12142#issuecomment-1204416869
        ctx.fillShape(&Shape{ .rect = .{ .w = w, .h = h } }, nvg.rgb(255, 255, 255));

        // TODO: maybe we don't need document here anymore and we could render layout tree
        ctx.renderNode(&layout_nodes[document.node.id]);
    }
};

const RenderContext = struct {
    vg: *nvg,

    const Self = @This();

    fn renderNode(self: *Self, node: *LayoutNode) void {
        const l = &node.layout;
        var rect = Rect{ .x = l.pos.x, .y = l.pos.y, .w = l.size.width, .h = l.size.height };

        if (node.text) |t| self.drawText(rect, t) else self.drawContainer(rect, &node.style, node.first_child);
    }

    fn drawContainer(self: *Self, rect: Rect, style: *const Style, first_child: ?*LayoutNode) void {
        // split open/close so we can skip invisibles AND we can also reduce stack usage per each recursion
        // TODO: @call(.{ .modifier = .never_inline }, ...)
        if (self.openContainer(rect, style)) {
            self.vg.translate(rect.x, rect.y);

            var next = first_child;
            while (next) |ch| : (next = ch.next) {
                self.renderNode(ch);
            }

            self.closeContainer();
        }
    }

    fn openContainer(self: *Self, rect: Rect, style: *const Style) bool {
        // we don't have to save/restore() if we can skip the whole subtree
        if (style.opacity == 0) {
            return false;
        }

        // restored later
        self.vg.save();

        if (rect.w == 0 or rect.h == 0) {
            // skip but recur
            return true;
        }

        if (style.opacity != 1.0) {
            const current = self.vg.ctx.getState().alpha;
            self.vg.globalAlpha(current * style.opacity);
        }

        const shape = if (!std.meta.eql(style.border_radius, .{ 0, 0, 0, 0 }))
            Shape{ .rrect = .{ .rect = rect, .radii = style.border_radius } }
        else
            Shape{ .rect = rect };

        // if let Some(matrix) = &style.transform {
        //     self.canvas.concat(matrix);
        // }

        // if (style.shadow) |shadow| {
        //     self.drawShadow(&shape, shadow);
        // }

        // if (style.outline) |outline| {
        //     self.drawOutline(&shape, outline);
        // }

        // if style.clip {
        //     self.clipShape(&shape, ClipOp::Intersect, true /*style.transform.is_some()*/);
        // }

        if (!std.meta.eql(style.background_color, TRANSPARENT)) {
            self.drawBgColor(&shape, style.background_color);
        }

        // TODO: image(s)

        // TODO: scroll
        // self.vg.translate(dx, dy);

        return true;
    }

    fn closeContainer(self: *Self) void {
        // TODO: optional border

        self.vg.restore();
    }

    fn drawBgColor(self: *Self, shape: *const Shape, color: Color) void {
        self.fillShape(shape, color);
    }

    fn fillShape(self: *Self, shape: *const Shape, color: Color) void {
        // std.debug.print("shape {any}\n", .{shape});

        self.vg.beginPath();

        switch (shape.*) {
            .rect => |rect| self.vg.rect(rect.x, rect.y, rect.w, rect.h),
            .rrect => |rrect| {
                const rect = &rrect.rect;
                // TODO: if (radii[0] == radii[1] ...) else ...
                self.vg.roundedRect(rect.x, rect.y, rect.w, rect.h, rrect.radii[0]);
            },
        }

        self.vg.fillColor(color);
        self.vg.fill();
    }

    fn drawText(self: *Self, rect: Rect, text: []const u8) void {
        self.vg.fontFace("sans");
        self.vg.fontSize(16);
        self.vg.fillColor(nvg.rgb(0, 0, 0));
        _ = self.vg.text(rect.x, rect.y + 16, text);
    }
};
